### PR TITLE
Fix bugs and refactor code

### LINKED ISSUE #78: .osu parser fails to parse beatmap correctly if whitespace between sections don't exist
Example case:
```
[Metadata]
Title:HONESTY
TitleUnicode:HONESTY
Artist:GYZE
ArtistUnicode:GYZE
Creator:Bibbity Bill
Version:StunterLetsPlay's Originality
Source:
Tags:
BeatmapID:1711767
BeatmapSetID:814850
[Difficulty]
HPDrainRate:4
CircleSize:4.2
OverallDifficulty:10
ApproachRate:9.8
SliderMultiplier:2
SliderTickRate:1
```
From https://osu.ppy.sh/beatmapsets/814850#osu/1711767

`Difficulty` section would be skipped due to `BufferedReader`'s behavior.

### PR DESCRIPTION
**Pull request by Rian8337**
- Fixed #78 
- Fixed a false NaN comparison
- Refactor many parts of code

### CODE REVIEW

**Comment by EdrowsLuo** [comment]
have you test it?

**Comment by Rian8337** [comment]
Yes, I've attempted to load the beatmap that is linked in #78 and it was able to load properly now.

My initial fix only considered whitespaces between beatmap sections while in fact it could be anywhere, hence why I made the "better" fix. Now cases like this should also be fixed:

```
[Metadata]
Title:HONESTY
TitleUnicode:HONESTY
Artist:GYZE
ArtistUnicode:GYZE
Creator:Bibbity Bill
Version:StunterLetsPlay's Originality
Source:
Tags:
BeatmapID:1711767
BeatmapSetID:814850

[Difficulty]
HPDrainRate:4
// whitespace in here
CircleSize:4.2
OverallDifficulty:10
ApproachRate:9.8
// another whitespace in here
SliderMultiplier:2
SliderTickRate:1
```

File: src/ru/nsu/ccfit/zuev/osu/OSUParser.java [hunk 1]
```diff
@@ -427,18 +348,32 @@ private String getNextSectionName() {
...
+                if (s.isEmpty()) {
+                    continue;
+                }
+
+                if (!s.matches(".+:.*")) {
                     if (s.matches("//.+")) {
                         continue;
                     } else {
+                        // If we don't meet a comment, it means we have arrived at a new section. In that case, go back to the previous line.
+                        // This ensures that reader.readLine() in getNextSectionName() returns the mentioned section.
```
Comment (EdrowsLuo) [inline src/ru/nsu/ccfit/zuev/osu/OSUParser.java:369]: There is a leap in logic, "don't meet a valid key-value pair or comment" don't always means a new section in logic.
DIrectly check if this line is "[SectionName]" would be better.  

And if you have enough time, you can try to reorganize the .osu file parsing logic here, there are still some bad designed code.

**Review by EdrowsLuo — CHANGES_REQUESTED** [review]

File: src/ru/nsu/ccfit/zuev/osu/OSUParser.java (hunk 1 above)
Comment (Rian8337) [reply src/ru/nsu/ccfit/zuev/osu/OSUParser.java:369]: Ah, sure.

**Comment by Rian8337** [comment]
Alright, I'm done. Please let me know if any improvements are needed.

**Comment by EdrowsLuo** [comment]
> Alright, I'm done. Please let me know if any improvements are needed.

Still bad code quality, :(
Can you tolerate a method with more than 200 lines code?

**Comment by Rian8337** [comment]
Separated `breakMetaData` to multiple methods and refactored more codes, should be better now in my opinion.

**Comment by Rian8337** [comment]
Attempted to load my whole songs library (16,000 beatmaps) with the latest commit, it went fine. Should be safe now, I can ask Discord users to test later.

**Pull request merged**